web/src/people.ts

```diff
@@ -1945,6 +1966,146 @@ export function populate_valid_user_ids(params: StateData["user_groups"]): void
+    });
+    fetch_users_storage.promise_for_in_transit.set(user_ids_pending_fetch, {
+        promise: promise_for_pending_fetch,
+        resolver: resolver_for_promise_for_pending_fetch,
+    });
+
+    fetch_users_storage.promise_for_pending = undefined;
+    fetch_users_storage.promise_resolver_for_pending = undefined;
+
+    void fetch_users(user_ids_pending_fetch).then(async (fetched_users) => {
+        for (const user of fetched_users) {
+            if (user.is_active) {
+                add_active_user(user);
+            } else {
+                non_active_user_dict.set(user.user_id, user);
+                _add_user(user);
+            }
+        }
+
+        // Resolve promises waiting on this fetch after updating the data locally.
+        fetch_users_storage.promise_for_in_transit.get(user_ids_pending_fetch)!.resolver();
+        // Clean up in transit promise for this fetch.
+        fetch_users_storage.promise_for_in_transit.delete(user_ids_pending_fetch);
+        // Remove fetched users from in transit user ids.
+        fetch_users_storage.in_transit_user_ids =
+            fetch_users_storage.in_transit_user_ids.difference(user_ids_pending_fetch);
+
+        await promise_for_user_ids_in_transit;
+        // Resolve promises waiting on the complete fetch.
+        fetch_users_storage.promise_for_in_transit.get(user_ids_to_fetch)!.resolver();
+    });
+}
+
+export async function fetch_users_from_ids_internal(user_ids: number[]): Promise<unknown> {
+    // NOTE: NEVER USE THIS FUNCTION DIRECTLY.
+    // Call get_or_fetch_users_from_ids instead.
+    const unknown_user_ids = new Set(
+        user_ids.filter((user_id) => !people_by_user_id_dict.has(user_id)),
+    );
+
+    // We already have data for all requested users.
+    if (unknown_user_ids.size === 0) {
+        return undefined;
+    }
+
+    // If the fetches in progress contain all of the unknown user IDs,
+    // return a promise that resolves when that fetch completes.
+    if (
+        unknown_user_ids.intersection(fetch_users_storage.in_transit_user_ids).size ===
+        unknown_user_ids.size
+    ) {
+        // Await for all the fetches in progress for the unknown user IDs.
+        const {promise_for_user_ids_in_transit, user_ids_pending_fetch} =
+            get_combined_promise_for_user_ids(unknown_user_ids);
+        assert(user_ids_pending_fetch.size === 0);
+        return promise_for_user_ids_in_transit;
+    }
+
+    // Add users to be fetched in the next fetch attempt.
+    fetch_users_storage.pending_user_ids =
+        fetch_users_storage.pending_user_ids.union(unknown_user_ids);
+
+    // Return promise for pending fetch if it exists.
+    if (fetch_users_storage.promise_for_pending !== undefined) {
+        return fetch_users_storage.promise_for_pending;
+    }
+
+    // Create promise for a next fetch attempt.
+    const promise = new Promise<void>((resolve) => {
+        fetch_users_storage.promise_resolver_for_pending = resolve;
```

Comment:
Why aren't we doing something with the `reject` function for this new promise? 

It certainly looks like the `fetch_users_from_server` code path doesn't retry indefinitely, and thus can end up getting an error due to say, a networking issue. I suspect the bug is that we do want this code path to do reasonable error handling, which probably does look like some version of retrying.